subsys/net/ip/utils.c

```diff
@@ -518,3 +518,230 @@ bool net_header_fits(struct net_pkt *pkt, u8_t *hdr, size_t hdr_size)
+{
+	unsigned long tmp;
+	char *endptr;
+
+	tmp = strtoul(buf, &endptr, 10);
+	if ((endptr == buf && tmp == 0) ||
+	    !(*buf != '\0' && *endptr == '\0') ||
+	    ((unsigned long)(unsigned short)tmp != tmp)) {
+		return false;
+	}
+
+	*port = tmp;
+
+	return true;
+}
+#endif /* CONFIG_NET_IPV6 || CONFIG_NET_IPV4 */
+
+bool net_ipaddr_parse(const char *str, size_t str_len, struct sockaddr *addr)
+{
+	int count, i, ret;
+	u16_t port;
+	char *ptr;
+
+	if (*str == '[') {
+#if defined(CONFIG_NET_IPV6)
+		/* IPv6 address with port number */
+		struct in6_addr *addr6;
+		char ipaddr[INET6_ADDRSTRLEN + 1];
+		int end, len;
+
+		ptr = memchr(str, ']', str_len);
+		if (!ptr) {
+			return false;
+		}
+
+		len = min(INET6_ADDRSTRLEN, str_len);
+		end = min(len, ptr - (str + 1));
+
+		memcpy(ipaddr, str + 1, end);
+		ipaddr[end] = '\0';
+
+		addr6 = &net_sin6(addr)->sin6_addr;
+
+		ret = net_addr_pton(AF_INET6, ipaddr, addr6);
+		if (ret < 0) {
+			return false;
+		}
+
+		net_sin6(addr)->sin6_family = AF_INET6;
+
+		if ((ptr + 1) < (str + str_len) && *(ptr + 1) == ':') {
+			len = str_len - end;
+
+			/* Re-use the ipaddr buf for port conversion */
+			memcpy(ipaddr, ptr + 2, len);
+			ipaddr[len] = '\0';
+
+			ret = convert_port(ipaddr, &port);
+			if (!ret) {
+				return false;
+			}
+
+			net_sin6(addr)->sin6_port = htons(port);
+
+			NET_DBG("IPv6 host %s port %d",
+				net_addr_ntop(AF_INET6,
+					      &net_sin6(addr)->sin6_addr,
+					      ipaddr, sizeof(ipaddr) - 1),
+				port);
+		} else {
+			NET_DBG("IPv6 host %s",
+				net_addr_ntop(AF_INET6,
+					      &net_sin6(addr)->sin6_addr,
+					      ipaddr, sizeof(ipaddr) - 1));
+		}
+
+		return true;
+#else
+		return false;
+#endif /* CONFIG_NET_IPV6 */
+	}
+
+	for (count = i = 0; str[i] && i < str_len; i++) {
+		if (str[i] == ':') {
+			count++;
+		}
+	}
+
+	if (count == 1) {
+#if defined(CONFIG_NET_IPV4)
+		/* IPv4 address with port number */
+		char ipaddr[NET_IPV4_ADDR_LEN + 1];
+		struct in_addr *addr4;
+		int end, len;
+
+		ptr = memchr(str, ':', str_len);
+		if (!ptr) {
+			return false;
+		}
+
+		len = min(NET_IPV4_ADDR_LEN, str_len);
+		end = min(len, ptr - str);
+
+		memcpy(ipaddr, str, end);
+		ipaddr[end] = '\0';
+
+		addr4 = &net_sin(addr)->sin_addr;
+
+		net_sin6(addr)->sin6_family = AF_INET;
+
+		ret = net_addr_pton(AF_INET, ipaddr, addr4);
+		if (ret < 0) {
+			return false;
+		}
+
+		memcpy(ipaddr, ptr + 1, str_len - end);
+		ipaddr[str_len - end] = '\0';
+
+		ret = convert_port(ipaddr, &port);
+		if (!ret) {
+			return false;
+		}
+
+		net_sin(addr)->sin_port = htons(port);
+
+		NET_DBG("IPv4 host %s port %d",
+			net_addr_ntop(AF_INET,
+				      &net_sin(addr)->sin_addr,
+				      ipaddr, sizeof(ipaddr) - 1),
+			port);
+		return true;
+#else
+		return false;
+#endif /* CONFIG_NET_IPV4 */
+	}
+
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV6)
+	char ipaddr[INET6_ADDRSTRLEN + 1];
+	int end, len;
+
+	len = min(NET_IPV4_ADDR_LEN, str_len);
+	end = min(len, strlen(str));
```

Comment:
This passed from the other review, but `strnlen()` should be used here, since `str` is not guaranteed to have a NUL terminator.  Other uses of `strlen()` here should be changed.  Apparently, minimal libc does not have a `strnlen()` implementation, but it's trivial to add one.